backend/api/src/supabase-search-contract.ts

```diff
@@ -0,0 +1,188 @@
+import { Contract } from 'common/contract'
+import { createSupabaseDirectClient } from 'shared/supabase/init'
+import { z } from 'zod'
+import { Json, MaybeAuthedEndpoint, validate } from './helpers'
+
+const bodySchema = z.object({
+  term: z.string(),
+  filter: z.string(),
+  sort: z.string(),
+  offset: z.number().gte(0),
+  limit: z.number().gt(0),
+  fuzzy: z.boolean().optional(),
+  groupId: z.string().optional(),
+  creatorId: z.string().optional(),
+})
+
+export const supabasesearchcontracts = MaybeAuthedEndpoint(
+  async (req, auth) => {
+    const { term, filter, sort, offset, limit, fuzzy, groupId, creatorId } =
+      validate(bodySchema, req.body)
+    const pg = createSupabaseDirectClient()
+    const searchMarketSQL = getSearchContractSQL({
+      term: term,
+      filter: filter,
+      sort: sort,
+      offset: offset,
+      limit: limit,
+      fuzzy: fuzzy,
+      groupId: groupId,
+      creatorId: creatorId,
+      uid: auth?.uid,
+    })
+    console.log(searchMarketSQL)
+    const contracts = await pg.map(
+      searchMarketSQL,
+      [],
+      (r) => r.data as Contract
+    )
+    return (contracts as unknown as Json) ?? ([] as unknown as Json)
+  }
+)
+
+function getSearchContractSQL(contractInput: {
+  term: string
+  filter: string
+  sort: string
+  offset: number
+  limit: number
+  fuzzy?: boolean
+  groupId?: string
+  creatorId?: string
+  uid?: string
+}) {
+  const { term, filter, sort, offset, limit, fuzzy, groupId, creatorId, uid } =
+    contractInput
+  let query = ''
+  const emptyTerm = term.length === 0
+  const whereSQL = getSearchContractWhereSQL(filter, sort, creatorId, uid)
+  if (groupId) {
+    if (emptyTerm) {
+      query = `
+        SELECT contractz.data
+        FROM (
+          select contracts.*, 
+          group_contracts.group_id 
+          from contracts 
+          join group_contracts 
+          on group_contracts.contract_id = contracts.id) 
+        as contractz
+        ${whereSQL}
+        AND contractz.group_id = '${groupId}'`
+    }
+    // if fuzzy search within group
+    else if (fuzzy) {
+      query = `
+        SELECT contractz.data
+        FROM (
+            SELECT contracts.*,
+                similarity(contracts.question, '${term}') AS similarity_score,
```

Comment:
This approach looks very susceptible to SQL injection. We either have to get a library to clean all the parameters (term, filter sort, etc) or somehow compose the queries with pg-promise.